### PR TITLE
🎨 Add "parallelism" parameter to "add_files" syscall and MigrateTable, SnapshotTable.

### PR DESCRIPTION
**Pull request by kingeasternsun**
…, SnapshotTable.

Signed-off-by: kingeasternsun <kingeasternsun@gmail.com>

Based on    `Speed up TableMigration By collect the DafaFile In parallel`   #3876 ,  
- add parallelism parameter to `add_files` and the related functions.
- for compatibility，I Keep the original method  and let original method call the newer method with default parallelism = 1
- add parallelism parameter to MigrateTableProcedure
- add parallelism parameter to SnapshotTableProcedure

### CODE REVIEW

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/SnapshotTableProcedure.java
Comment (amogh-jahagirdar) [inline spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/SnapshotTableProcedure.java:102]: Nit: I think it's a bit cleaner to use ternary here

`int parallelism = args.getInt(4) != null ? args.getInt(4) : 1`

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/MigrateTableProcedure.java
Comment (amogh-jahagirdar) [inline spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/MigrateTableProcedure.java:108]: Same nit as below i think ternary assignment works well here

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/MigrateTableProcedure.java
Comment (amogh-jahagirdar) [inline spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/MigrateTableProcedure.java:43]: Any reason we can't just call this parameter "parallelism" ? I think it will only be data files involved

**Comment by amogh-jahagirdar** [comment]
Left a review, thanks for the contribution @kingeasternsun !

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/MigrateTableProcedure.java
Comment (kingeasternsun) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/MigrateTableProcedure.java:43]: Thanks for your reviews, It because this  https://github.com/apache/iceberg/pull/3973#discussion_r794150006

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/SnapshotTableProcedure.java
Comment (kingeasternsun) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/SnapshotTableProcedure.java:102]: Thanks for your advice, I will fix this.

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/MigrateTableProcedure.java
Comment (kingeasternsun) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/MigrateTableProcedure.java:108]: Thanks for your advice, I will fix this.

**Comment by kingeasternsun** [comment]
> Left a review, thanks for the contribution @kingeasternsun ! Also looks like spotless checks are failing which you can fix by running `./gradlew :iceberg-api:spotlessJavaCheck`

Thanks for your advice, I have fixed them. Will you merge this please.

**Comment by amogh-jahagirdar** [comment]
@kingeasternsun A maintainer should take a look when they get a chance. @RussellSpitzer @aokolnychyi @szehon-ho @jackye1995  when you get a chance could you take a look? Thanks!

File: api/src/main/java/org/apache/iceberg/actions/MigrateTable.java
Comment (jackye1995) [inline api/src/main/java/org/apache/iceberg/actions/MigrateTable.java:58]: I think we can default to 1?

File: api/src/main/java/org/apache/iceberg/actions/MigrateTable.java
Comment (jackye1995) [inline api/src/main/java/org/apache/iceberg/actions/MigrateTable.java:57]: can we simplify and unify all the naming? We can just call it `parallelism` and use this name across the board. It also matches the naming in `TableMigrationUtil`

Comment (kingeasternsun) [reply api/src/main/java/org/apache/iceberg/actions/MigrateTable.java:57]: /LGTM。  @RussellSpitzer @aokolnychyi @szehon-ho @jackye1995 when you get a chance could you please take a look? Thanks。

Comment (jackye1995) [reply api/src/main/java/org/apache/iceberg/actions/MigrateTable.java:57]: Thanks, I see you changed some places, there are 2 places I think can be also consistent, the first is this place, where we can just do `default MigrateTable withParallelism(int parallelism)` instead of using a different name and introduce a `reader` concept.

The other place is in the procedure, where you have `        ProcedureParameter.optional("max_concurrent_read_datafiles", DataTypes.IntegerType)
`, we can just name the option `parallelism` as well.

What do you think?

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/SnapshotTableProcedure.java
Comment (kingeasternsun) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/SnapshotTableProcedure.java]: > Nit: I think it's a bit cleaner to use ternary here
> 
> `int parallelism = args.getInt(4) != null ? args.getInt(4) : 1`

Thanks, I have fixed.

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/MigrateTableProcedure.java
Comment (kingeasternsun) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/MigrateTableProcedure.java]: > Same nit as below i think ternary assignment works well here

Thanks, I have fixed @amogh-jahagirdar .

File: api/src/main/java/org/apache/iceberg/actions/MigrateTable.java
Comment (kingeasternsun) [reply api/src/main/java/org/apache/iceberg/actions/MigrateTable.java:58]: > I think we can default to 1?

/LGTM,  But I'm not familiar with JAVA,  Could you give more detail about how set this default to 1 , many thanks. @jackye1995

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/MigrateTableProcedure.java
Comment (kingeasternsun) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/MigrateTableProcedure.java:43]: @amogh-jahagirdar Another reason is that `importSparkPartitions`  https://github.com/kingeasternsun/iceberg/blob/feature/add-parallelism-add-files/spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkTableUtil.java#L620-L630 has a local variable named `parallelism`

File: api/src/main/java/org/apache/iceberg/actions/MigrateTable.java
Comment (kingeasternsun) [reply api/src/main/java/org/apache/iceberg/actions/MigrateTable.java:57]: Thanks for your advice,It looks good to me,  but in function `importSparkPartitions` https://github.com/kingeasternsun/iceberg/blob/feature/add-parallelism-add-files/spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkTableUtil.java#L620-L630 already has a local variable named `parallelism`

File: api/src/main/java/org/apache/iceberg/actions/MigrateTable.java
Comment (RussellSpitzer) [reply api/src/main/java/org/apache/iceberg/actions/MigrateTable.java:58]: I think unsupported is right here, since we are basically saying that the method is unsupported not that there isn't a default.

Comment (RussellSpitzer) [reply api/src/main/java/org/apache/iceberg/actions/MigrateTable.java:58]: Also I think our default should probably be > 1

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkTableUtil.java
Comment (RussellSpitzer) [inline spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkTableUtil.java:410]: Unfortunately this is a public api, We should probably add a new signature and deprecate the old one, or keep the old one. I think in this case keeping the old one without deprecation is fine but we probably should not break binary compatibility unless we have to.

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkTableUtil.java
Comment (RussellSpitzer) [inline spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkTableUtil.java:489]: can we set the default here to our worker pool default thread count?
https://github.com/kingeasternsun/iceberg/blob/feature/add-parallelism-add-files/core/src/main/java/org/apache/iceberg/util/ThreadPools.java#L38

I think this is probably safe

File: api/src/main/java/org/apache/iceberg/actions/MigrateTable.java
Comment (kingeasternsun) [reply api/src/main/java/org/apache/iceberg/actions/MigrateTable.java:58]: Thanks @RussellSpitzer for your reviews again, So I just Keep it as it is using unsupported.

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkTableUtil.java
Comment (kingeasternsun) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkTableUtil.java:410]: Thanks again for your reviews, this PR's changes is almost same with the older #3973 :)

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkTableUtil.java
Comment (kingeasternsun) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkTableUtil.java:489]: /LGTM.
I concern that if we set the default here to our worker pool default thread count, whether this change violates Compatibility, because older version is single threaded.

Comment (RussellSpitzer) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkTableUtil.java:489]: This is just a performance chance, unless we think this will break somehow we can change the implementation of apis, just not apis or their outputs.

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/AddFilesProcedure.java
Comment (RussellSpitzer) [inline spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/AddFilesProcedure.java:123]: I'm not sure how we ended up with this pattern here, in the other procedures we do
```java
var = (arg.nullat(?)) default : arg.get(?);
```
Although There is probably no reason to change this now. I do think a default of 1 is still probably too low. I feel like this is a `gotcha` default and is basically never correct.

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkTableUtil.java
Comment (kingeasternsun) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkTableUtil.java:489]: Ok, I'll fixed it.

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/AddFilesProcedure.java
Comment (kingeasternsun) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/AddFilesProcedure.java:123]: So we set default here to worker pool default thread count too?

File: api/src/main/java/org/apache/iceberg/actions/MigrateTable.java
Comment (kingeasternsun) [reply api/src/main/java/org/apache/iceberg/actions/MigrateTable.java:57]: @jackye1995  Ok, I have fixed it.

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/MigrateTableProcedure.java
Comment (kingeasternsun) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/MigrateTableProcedure.java]: @amogh-jahagirdar ,Thanks for your review,  I have fixed it.

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkTableUtil.java
Comment (kingeasternsun) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkTableUtil.java]: @RussellSpitzer , I have fixed it.

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/AddFilesProcedure.java
Comment (kingeasternsun) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/AddFilesProcedure.java]: @RussellSpitzer , Thanks for your review, I have fixed it to `ThreadPools.WORKER_THREAD_POOL_SIZE`

**Comment by kingeasternsun** [comment]
Hi, should I squash the commits before merge @RussellSpitzer @jackye1995 @amogh-jahagirdar ?

**Comment by sweetpythoncode** [comment]
@RussellSpitzer Can you take a look at it when u have time, the current method is super slow on a big number of partitions, please

**Review by amogh-jahagirdar — APPROVED** [review]
This looks good to me thanks @kingeasternsun ! I see we're missing benchmarks for `AddFilesProcedure` and the others, I think it's worth having JMH benchmarks for these. I'll create a separate issue for that though.

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/actions/MigrateTableSparkAction.java
Comment (amogh-jahagirdar) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/actions/MigrateTableSparkAction.java:62]: Oh good catch, yeah it seems like the worker pool size is used in the procedures but not in the action. That should be fixed

**Comment by jackye1995** [comment]
This PR looks ready to go, @kingeasternsun could you rebase it?

**Comment by jackye1995** [comment]
Looks like there is no update, @kingeasternsun do you still plan to move this forward? If not we can open a new PR and mark you as coauthor.

**Comment by hazelnutsgz** [comment]
I feel like the number of concurrent thread should be somehow determined by the number of files in that partition. Let's say partition A got 10000 files and partition B got 10 files. The number of threads should be `

``` number of files in the partition * configurable_factor```, where configurable_factor is between 0 to 1.

File: api/src/main/java/org/apache/iceberg/actions/MigrateTable.java
Comment (kingeasternsun) [reply api/src/main/java/org/apache/iceberg/actions/MigrateTable.java]: Thank very much for your reviews  @jackye1995 , but for some reason I am not quite able to continue updating this MR In time, So I agree with your suggestion that “open a new PR” . Best wishes.

**Comment by kingeasternsun** [comment]
@jackye1995 Thanks for your review. 

>  If not we can open a new PR and mark you as coauthor.

Yes, I agree with that

**Pull request closed without merging**
